src/util/PendlePTHelper.sol

```diff
@@ -0,0 +1,281 @@
+
+        _revertIfMarketNotSet(market);
+
+        IERC20 pt = IERC20(markets[market].pt);
+        IERC20 yt = IERC20(markets[market].yt);
+        DOLA.safeTransferFrom(msg.sender, address(this), amount);
+        DOLA.approve(router, amount);
+        // Avoid accounting for possibly stucked token for previous bad input and allow recovery
+        uint256 ptBalBefore = pt.balanceOf(address(this));
+        uint256 ytBalBefore = yt.balanceOf(address(this));
+        (bool success, ) = router.call(callData);
+        if (!success) revert PendleSwapFailed();
+
+        uint256 ptBal = pt.balanceOf(address(this)) - ptBalBefore;
+
+        if (ptBal < minMint) revert InsufficientPT();
+        if (recipient != address(this)) pt.safeTransfer(recipient, ptBal);
+        // Send YT to user if specified
+        if (ytRecipient != address(0)) {
+            uint256 ytBalMinted = yt.balanceOf(address(this)) - ytBalBefore;
+            if (ytBalMinted < minMint) revert InsufficientYT();
+            yt.safeTransfer(ytRecipient, ytBalMinted);
+        }
+
+        return ptBal;
+    }
+
+    /**
+     * @notice Redeems PT token for DOLA.
+     * @dev Used by the ALE but can be called by anyone. Carefully review input data for Pendle API.
+     * The receiver in Pendle API has to be set same as the recipient. If a REDEEM is performed, include a ytProvider with enough allowance.
+     * @param amount The amount of PT token to be redeemed (and YT if specified).
+     * @param data Encoded address of the market, minimum amount of DOLA to receive, ytProvider, and Pendle callData.
+     * @return dolaAmount The amount of DOLA redeemed.
+     */
+    function transformFromCollateral(
+        uint256 amount,
+        bytes calldata data
+    ) external override returns (uint256 dolaAmount) {
+        dolaAmount = transformFromCollateral(amount, msg.sender, data);
+    }
+
+    /**
+     * @notice Redeems Collateral for DOLA.
+     * @dev The receiver in Pendle API has to be set same as the recipient. If a REDEEM is performed, include a ytProvider with enough allowance.
+     * @param amount The amount of PT Token to be redeemed (and YT if specified).
+     * @param recipient The address to which the underlying token is transferred.
+     * @param data Encoded address of the market, minimum amount of DOLA to receive for the recipient, ytProvider, and Pendle callData.
+     * @return dolaAmount The amount of DOLA redeemed.
+     */
+    function transformFromCollateral(
+        uint256 amount,
+        address recipient,
+        bytes calldata data
+    ) public override returns (uint256 dolaAmount) {
+        (
+            address market,
+            uint256 minOut, // Minimum amount of DOLA to receive
+            address ytProvider,
+            bytes memory callData
+        ) = abi.decode(data, (address, uint256, address, bytes));
+        _revertIfMarketNotSet(market);
+
+        if (ytProvider != address(0)) {
```

Comment:
Should have a check and error message in case `ytProvider` doesn't have a large enough balance to supply the YT tokens.

Reply:
Thinking about it, how is the `ytProvider` supposed to work? If it's a user address or contract, that's a security vulnerability.
1. Honest user approves YT tokens as `ytProvider` for the helper contract
2. Attacker spots approve in mempool.
3. Attacker deploys PT collateral to market, then withdraws using `transformFromCollateral` with honest user as `ytProvider`

Reply:
Either only allow `ytProvider` and recipient to be same address, or only allow `ytProvider` to be `msg.sender`.